lib/src/space/grounding.rs

```diff
@@ -58,24 +58,16 @@ pub(crate) fn atom_to_trie_key(atom: &Atom) -> TrieKey<SymbolAtom> {
                 expr.children().iter().for_each(|child| fill_key(child, tokens));
                 tokens.push(TrieToken::RightPar);
             },
-            // FIXME, see below
-            Atom::Grounded(g) => {
+            Atom::Grounded(g) if g.as_grounded().as_match().is_none() => {
+                // TODO: Adding Hash on grounded atoms matched by equality is
+                // required in order to make TrieToken::Exact be generated for
+                // them.
                 let mut h = DefaultHasher::new();
                 match (*g).serialize(&mut h) {
                     Ok(()) => { tokens.push(TrieToken::Exact(SymbolAtom::new(ImmutableString::Allocated(h.finish().to_string())))) }
                     Err(_) => { tokens.push(TrieToken::Wildcard) }
```

Comment:
Should all (grounded) atoms that do not provide a custom match, provide a serializer?
I.e. as_match would conceptually be `Result<CustomMatch, Serializer>`.
This way HE can get rid of the linear search disambiguation logic, because a grounded item either fits into the default matcher, or it provides its own matching logic.

Reply:
Not 100% sure I got you right. If custom matching logic is implemented for the grounded atom then in general we cannot use serialization to get `TrieToken::Exact` anyway, so probably you didn't mean this. Regarding `Serializer` for the default matching logic I am going to add this in next PR. In this PR my goal was to patch as quick as possible the behavior for the case when custom matching logic and `serialize` are implemented by the grounded atom at the same time.

Reply:
This PR LGTM, my analysis was that there are four cases:
CustomMatch, Serialize => should match using custom logic
Serialize => should match using serialized version of the atom wrapped in exact
CustomMatch => should match using custom logic
=> should never happen, let's make this illegal and unrepresentable

Reply:
Ah, ok, so you mean `Trie` implementation could call custom matching logic when matching key. Yes it is possible to implement it. The reason it is not implemented yet is that I wanted to have the single instance of the matching procedure instead of two instances (searching in space and matching with a single atom). 

It is indeed possible to implement matching a single atom via matching with space which contains a single atom and thus still have single instance of the matching procedure. Or for instance parameterize matching procedure by a function which provides next atom to match, then implement it differently for a single atom case and for `Trie` search case.

I think it makes sense and I think I will try to implement this.